### PR TITLE
Force default locale for cache model write/read

### PR DESCRIPTION
**Pull request by zhaixuejun1993**
### Details:
 - Check the value of setlocale for export/import, if different with "C" will set to "C" and record the original value, after export/import done will reset to the original.
 - *Fix the error caused by pugixml library with The setlocale function installs the specified system locale or its portion as the new C locale. different C may return unexpected results with setlocal()*

### Tickets:
 - https://github.com/openvinotoolkit/openvino/issues/24370

### CODE REVIEW

**Comment by riverlijunjie** [comment]
@ilya-lavrenov could you check whether this fixing is reasonable? Thanks!

File: src/inference/src/cache_manager.hpp [hunk 1]
```diff
@@ -99,16 +99,32 @@ class FileStorageCacheManager final : public ICacheManager {
...
+        if (std::strcmp(original_plocal, "C") != 0) {
+            setlocale(LC_ALL, "C");
+        }
         auto blobFileName = getBlobFile(id);
         if (ov::util::file_exists(blobFileName)) {
             std::ifstream stream(blobFileName, std::ios_base::binary);
             reader(stream);
         }
+        if (std::strcmp(original_plocal, "C") != 0) {
+            setlocale(LC_ALL, original_plocal);
```
Comment (ilya-lavrenov) [inline src/inference/src/cache_manager.hpp:126]: can we add tests for it?

File: src/inference/src/cache_manager.hpp [hunk 2]
```diff
@@ -99,16 +99,32 @@ class FileStorageCacheManager final : public ICacheManager {
 
 private:
     void write_cache_entry(const std::string& id, StreamWriter writer) override {
+        // Fix the bug caused by pugixml, which may return unexpected results if the locale is different from "C".
+        auto original_plocal = setlocale(LC_NUMERIC, nullptr);
+        if (std::strcmp(original_plocal, "C") != 0) {
```
Comment (ilya-lavrenov) [inline src/inference/src/cache_manager.hpp:104]: I suppose we can change it unconditionally ? anyway, we restore it after

Comment (ilya-lavrenov) [reply src/inference/src/cache_manager.hpp:104]: BTW, do we need the same for IR reader / serializer? It also operates with numbers

File: src/inference/src/cache_manager.hpp [hunk 3]
```diff
@@ -99,16 +99,32 @@ class FileStorageCacheManager final : public ICacheManager {
...
         std::ofstream stream(getBlobFile(id), std::ios_base::binary | std::ofstream::out);
         writer(stream);
+        if (std::strcmp(original_plocal, "C") != 0) {
+            setlocale(LC_ALL, original_plocal);
+        }
     }
 
     void read_cache_entry(const std::string& id, StreamReader reader) override {
+        // Fix the bug caused by pugixml, which may return unexpected results if the locale is different from "C".
+        auto original_plocal = setlocale(LC_NUMERIC, nullptr);
```
Comment (RyanMetcalfeInt8) [inline src/inference/src/cache_manager.hpp:116]: Something to be careful about here -- the char* that setlocale is returning here may be a reference to the current, globally set locale. So, by calling setlocale below, it be changing the string that 'original_plocal' is pointing to. 

Just a disclaimer... I read something like that from some stack overflow thread -- I don't know if it's true :smile:

File: src/inference/src/cache_manager.hpp (hunk 2 above)
Comment (zhaixuejun1993) [reply src/inference/src/cache_manager.hpp]: the reader / serializer is called by write_cache_entry() & read_cache_entry(), right?
If yes, this setlocale(LC_ALL, "C"); will take effect for them.

**Comment by ilya-lavrenov** [comment]
@RyanMetcalfeInt8 
Could you please check whether the fix helps in your case?

File: src/inference/src/cache_manager.hpp [hunk 4]
```diff
@@ -18,6 +18,25 @@
...
+    ScopedLocale(int category, std::string newLocale) : m_Category(category) {
+        m_OldLocale = setlocale(category, nullptr);
+        setlocale(m_Category, newLocale.c_str());
+    }
+    ~ScopedLocale() {
+        setlocale(m_Category, m_OldLocale.c_str());
+    }
+
+private:
+    int m_Category;
```
Comment (riverlijunjie) [inline src/inference/src/cache_manager.hpp:36]:
```suggestion
    int m_category;
```

File: src/inference/tests/functional/local_test.cpp [hunk 5]
```diff
@@ -286,3 +289,67 @@ TEST_F(LocaleTests, DISABLED_WithUSLocaleCPP) {
...
+            auto tensor = ov::test::utils::create_and_fill_tensor_normal_distribution(input.get_element_type(),
+                                                                                      input.get_shape(),
+                                                                                      0.0f,
+                                                                                      0.2f,
+                                                                                      7235346);
+            inputs.insert({input, tensor});
+        }
+
+        ov::Core core;
+        core.set_property(ov::cache_dir(cache_dir));
```
Comment (riverlijunjie) [inline src/inference/tests/functional/local_test.cpp:322]: Why not put `ov::cache_dir` into `compile_model`?

File: src/inference/tests/functional/local_test.cpp [hunk 6]
```diff
@@ -286,3 +289,67 @@ TEST_F(LocaleTests, DISABLED_WithUSLocaleCPP) {
...
+
+        auto output_from_model_read = getOutputBlob();
+        auto output_from_model_cached = getOutputBlob();
+
+        ov::test::utils::compare(output_from_model_read, output_from_model_cached);
+        ov::test::utils::removeFilesWithExt(cache_dir, "blob");
+    }
+};
+
+#if defined(ENABLE_OV_IR_FRONTEND)
```
Comment (riverlijunjie) [inline src/inference/tests/functional/local_test.cpp:345]: I supposed we should delete this line and merge them together.

**Comment by RyanMetcalfeInt8** [comment]
> @RyanMetcalfeInt8 Could you please check whether the fix helps in your case?

Do I need to build OpenVINO in order to try this, or can I grab a CI build?

**Comment by zhaixuejun1993** [comment]
> > @RyanMetcalfeInt8 Could you please check whether the fix helps in your case?
> 
> Do I need to build OpenVINO in order to try this, or can I grab a CI build?

The CI build is enough, I think.

**Comment by RyanMetcalfeInt8** [comment]
> > > @RyanMetcalfeInt8 Could you please check whether the fix helps in your case?
> > 
> > 
> > Do I need to build OpenVINO in order to try this, or can I grab a CI build?
> 
> The CI build is enough, I think.

Can you point me to where I can download the CI build? I was trying to find a link to artifacts from GitHub actions, but I couldn't seem to find it..

**Comment by zhaixuejun1993** [comment]
> > > > @RyanMetcalfeInt8 Could you please check whether the fix helps in your case?
> > > 
> > > 
> > > Do I need to build OpenVINO in order to try this, or can I grab a CI build?
> > 
> > 
> > The CI build is enough, I think.
> 
> Can you point me to where I can download the CI build? I was trying to find a link to artifacts from GitHub actions, but I couldn't seem to find it..

http://ov-share-03.iotg.sclab.intel.com/volatile/openvino_ci/private_builds/dldt/master/pre_commit/0924c59be2edba83c088890a5a03e5b3712b2e15/private_linux_ubuntu_20_04_release/cpack/

**Comment by ilya-lavrenov** [comment]
> > > > @RyanMetcalfeInt8 Could you please check whether the fix helps in your case?
> > > 
> > > 
> > > Do I need to build OpenVINO in order to try this, or can I grab a CI build?
> > 
> > 
> > The CI build is enough, I think.
> 
> Can you point me to where I can download the CI build? I was trying to find a link to artifacts from GitHub actions, but I couldn't seem to find it..

image

**Comment by RyanMetcalfeInt8** [comment]
> > > > > @RyanMetcalfeInt8 Could you please check whether the fix helps in your case?
> > > > 
> > > > 
> > > > Do I need to build OpenVINO in order to try this, or can I grab a CI build?
> > > 
> > > 
> > > The CI build is enough, I think.
> > 
> > 
> > Can you point me to where I can download the CI build? I was trying to find a link to artifacts from GitHub actions, but I couldn't seem to find it..
> 
> image

Thanks, I have verified that this PR resolves the locale issue.

File: src/inference/src/cache_manager.hpp [hunk 7]
```diff
@@ -18,6 +18,25 @@
 
 namespace ov {
 
+/**
+ * @brief This class limit the local env to a spicial value in sub-scope
```
Comment (riverlijunjie) [inline src/inference/src/cache_manager.hpp:22]:
```suggestion
 * @brief This class limits the locale env to a special value in sub-scope
```

**Review by ilya-lavrenov — APPROVED** [review]

**Review by riverlijunjie — APPROVED** [review]

**Pull request merged**